src/java.net.http/share/classes/jdk/internal/net/http/websocket/Frame.java

```diff
@@ -215,7 +215,10 @@ HeaderWriter fin(boolean value) {
             if (value) {
                 firstChar |=  0b10000000_00000000;
             } else {
-                firstChar &= ~0b10000000_00000000;
+                // The negation "~" sets the high order bits
+                // so the value is more than 16 bits and the
+                // compiler will emit a warning if not cast
+                firstChar &= (char) ~0b10000000_00000000;
             }
```

Comment:
Am I right in believing that there was an implicit cast to char here before and the only change is to make it explicit? ie. there is no actual behavior change?

Reply:
Yes - that's my understanding.

Reply:
These methods either set or clear a single bit in `firstChar`.
The constant is an `int` so its complement is a 32 bit int.

It could also be written as `~ (char) 0b1000000_0000000`.   Then the 16 bit unsigned char would be complemented.
Either way, the cast is needed to be explicit about the size of the constant.

Another way to avoid the cast would be to define the bit positions as:

`private static final char FIN_BIT = 0b10000000_00000000;
... etc.
`
Then the code in the methods would not need the cast.

```
    if (value) { 
        firstChar |= FIN_BIT;
    } else {
        firstChar &= ~FIN_BIT;
    }
```

Reply:
Ah! Good point. Maybe I should use a constant and get rid of the cast.

Reply:
👍 I'd put `_MASK` in the name along with whatever name is used for the bits in the frame spec .

Reply:
Done

Reply:
@RogerRiggs Actually - I'm no longer sure that this will work:

```
 char x = 0b10000000_0000000
x ==> '䀀'

jshell> var y = ~x
y ==> -16385

jshell> char y = ~x
|  Error:
|  incompatible types: possible lossy conversion from int to char
|  char y = ~x;
|           ^^
```

Reply:
So if x is a char, ~x seems to be an int :-(

Reply:
I have reverted adding constant fields. Too bad - the casts are back.

Reply:
Phooey:  Sorry to have lead you astray.  :(
And casting before `~` does the same not wanted result.